sw/airborne/modules/sensors/mag_qmc5883.c

```diff
@@ -0,0 +1,288 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with paparazzi; see the file COPYING.  If not, write to
+ * the Free Software Foundation, 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ */
+
+/**
+ * @file modules/sensors/mag_qmc5883.c
+ *
+ * Module for QST QMC5883 magnetometer, the DB vesion
+ */
+
+#include "modules/sensors/mag_qmc5883.h"
+#include "modules/sensors/mag_qmc5883_regs.h"
+#include "mcu_periph/uart.h"
+#include "pprzlink/messages.h"
+#include "modules/datalink/downlink.h"
+#include "generated/airframe.h"
+#include "modules/core/abi.h"
+
+#if PERIODIC_TELEMETRY
+#include "modules/datalink/telemetry.h"
+#include "pprzlink/messages.h"
+#endif
+
+#if MODULE_QMC5883_UPDATE_AHRS
+#include "modules/imu/imu.h"
+#include "modules/core/abi.h"
+
+#ifndef QMC5883_CHAN_X
+#define QMC5883_CHAN_X 0
+#endif
+#ifndef QMC5883_CHAN_Y
+#define QMC5883_CHAN_Y 1
+#endif
+#ifndef QMC5883_CHAN_Z
+#define QMC5883_CHAN_Z 2
+#endif
+#ifndef QMC5883_CHAN_X_SIGN
+#define QMC5883_CHAN_X_SIGN +
+#endif
+#ifndef QMC5883_CHAN_Y_SIGN
+#define QMC5883_CHAN_Y_SIGN +
+#endif
+#ifndef QMC5883_CHAN_Z_SIGN
+#define QMC5883_CHAN_Z_SIGN +
+#endif
+
+#if defined QMC5883_MAG_TO_IMU_PHI && defined QMC5883_MAG_TO_IMU_THETA && defined QMC5883_MAG_TO_IMU_PSI
+#define USE_MAG_TO_IMU 1
+static struct Int32RMat mag_to_imu; // rotation from mag to imu frame
+#else
+#define USE_MAG_TO_IMU 0
+#endif
+#endif
+
+struct Qmc5883 qmc;
+struct Int32Vect3 mag;
+
+#if PERIODIC_TELEMETRY
+
+static void mag_qmc5883_send_imu_mag_raw(struct transport_tx *trans, struct link_device *dev)
+{
+  pprz_msg_send_IMU_MAG_RAW(trans, dev, AC_ID,
+                            &mag.x,
+                            &mag.y,
+                            &mag.z);
+}
+
+static void mag_qmc5883_send_debug(struct transport_tx *trans, struct link_device *dev)
+{
+  pprz_msg_send_QMC5883_DEBUG(trans, dev, AC_ID,
+                              &qmc.status_debug,
+                              &qmc.trans_status,
+                              &qmc.raw_mag_data[0],
+                              &qmc.raw_mag_data[1],
+                              &qmc.raw_mag_data[2],
+                              &qmc.raw_mag_data[3],
+                              &qmc.raw_mag_data[4],
+                              &qmc.raw_mag_data[5],
+                              &qmc.mag_data[0],
+                              &qmc.mag_data[1],
+                              &qmc.mag_data[2],
+                              &qmc.status_register.DRDY,
+                              &qmc.status_register.OVL,
+                              &qmc.status_register.DOR);
+}
+
+#endif
+
+/**
+ * Initialization Function
+ */
+void mag_qmc5883_init(void)
+{
+  qmc.trans.status = I2CTransDone;
+  qmc.waddr = QMC5883_WRITE_ADDR;
+  qmc.raddr = QMC5883_READ_ADDR;
```

Comment:
you don't need 2 different addresses, only one on 8 bits, without the terminal 0/1. So in you case the `QMC5883_WRITE_ADDR` the R/W bit is added by the driver if needed.